src/geometryRTheta/time_solver/bsl_predcorr.hpp

```diff
@@ -123,74 +123,87 @@ class BslPredCorrRTheta : public ITimeSolverRTheta
-                };
-
-        RK2<host_t<DFieldMemRTheta>,
-            host_t<DVectorFieldMemRTheta<X, Y>>,
-            Kokkos::DefaultHostExecutionSpace>
+                .with("electrical_potential", electrical_potential_host);
+
+        // RK2 methods
+        std::function<void(DVectorFieldRTheta<X, Y>, DConstFieldRTheta)> define_advection_field
+                = [&](DVectorFieldRTheta<X, Y> advection_field, DConstFieldRTheta density) {
+                      ddc::parallel_deepcopy(density_host, density);
+                      // --- compute electrostatic potential:
+                      m_builder(get_field(density_coef_host), get_const_field(density_host));
+                      m_poisson_solver(charge_density, electrostatic_potential_coef_host);
+
+                      auto advection_field_host = ddcHelper::create_mirror_view_and_copy(
+                              Kokkos::DefaultHostExecutionSpace(),
+                              advection_field);
+
+                      // --- compute advection field:
+                      advection_field_computer(
+                              electrostatic_potential_coef_host,
+                              get_field(advection_field_host));
+
+                      ddcHelper::deepcopy(advection_field, advection_field_host);
+                  };
+
+        std::function<void(DFieldRTheta, DConstVectorFieldRTheta<X, Y>, double)> advect_density
+                = [&](DFieldRTheta density,
+                      DConstVectorFieldRTheta<X, Y> advection_field,
+                      double dt) {
+                      m_advection_solver(get_field(density), get_const_field(advection_field), dt);
+                  };
+
+        RK2<DFieldMemRTheta, DVectorFieldMemRTheta<X, Y>, Kokkos::DefaultExecutionSpace>
                 time_stepper(grid);
-        DFieldMemRTheta electrical_potential(grid);
-        host_t<DFieldMemRTheta> electrical_potential_host(grid);
+
+        // Iteration loop
         start_time = std::chrono::system_clock::now();
         for (int iter(0); iter < steps; ++iter) {
+            ddc::parallel_deepcopy(density, get_const_field(density_host));
             time_stepper
-                    .update(Kokkos::DefaultHostExecutionSpace(),
-                            density_host,
+                    .update(Kokkos::DefaultExecutionSpace(),
+                            density,
                             dt,
                             define_advection_field,
                             advect_density);
 
-
-            m_builder(get_field(density_coef), get_const_field(density_host));
-            m_poisson_solver(charge_density_coord, get_field(electrical_potential));
-            ddc::parallel_deepcopy(electrical_potential_host, electrical_potential);
+            ddc::parallel_deepcopy(density_host, get_const_field(density));
+            m_builder(get_field(density_coef_host), get_const_field(density_host));
```

Comment:
```suggestion
            m_builder(density_coef_host, get_const_field(density_host));
```
?

or change for `density_coef_alloc_host`.

Reply:
[c246b3e](https://github.com/gyselax/gyselalibxx/pull/395/commits/c246b3e9a7accf6fda3a0a7e714d52383fdceaab)